test/Npgsql.Tests/ConnectionTests.cs

```diff
@@ -1453,6 +1454,26 @@ public void PasswordSourcePrecendence()
             };
         }
 
+#if !NET461 // .NET 4.6.1 doesn't support cancellation tokens on socket operations, so no async timeouts
+        [Test, Description("Simulates a timeout during the authentication phase")]
+        [IssueLink("https://github.com/npgsql/npgsql/issues/3227")]
+        [Timeout(10000)]
+        public async Task TimeoutDuringAuthentication()
+        {
+            var builder = new NpgsqlConnectionStringBuilder(ConnectionString) { Timeout = 1 };
+            await using var postmasterMock = new PgPostmasterMock(builder.ConnectionString);
+            using var _ = CreateTempPool(postmasterMock.ConnectionString, out var connectionString);
+
+            var __ = postmasterMock.AcceptServer();
```

Comment:
```suggestion
            await postmasterMock.AcceptServer();
```

Reply:
But if we await here, we hang forever because we never open the connection below, no? That's the slight difficulty I was referring to above - as we're setting up both the server and client, either accepting or connecting has to happen in the background/in parallel.

Reply:
Yeah, my mistake - nvm then